lib/linked_list.rb

```diff
@@ -12,120 +12,329 @@ def initialize(value, next_node = nil)
 
 # Defines the singly linked list
 class LinkedList
-    def initialize
-      @head = nil # keep the head private. Not accessible outside this class
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
+  
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  def add_first(value)
+    current = @head
+    @head = Node.new(value, current)
+  end
+  
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  def search(value)
+    return false if @head == nil
+    current = @head
+    
+    while current do
+      if current.data == value
+        return true
+      else
+        current = current.next
+      end
     end
-
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    def add_first(value)
-      raise NotImplementedError
+    
+    return false
+  end
+  
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  def find_max
+    return nil if @head == nil
+    current = @head
+    max = @head.data
+    
+    while current do
+      if current.data > max
+        max = current.data
+      end
+      current = current.next
     end
-
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    def search(value)
-      raise NotImplementedError
+    
+    return max
+  end
+  
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  def find_min
+    return nil if @head == nil
+    current = @head
+    min = @head.data
+    
+    while current do
+      if current.data < min
+        min = current.data
+      end
+      current = current.next
     end
-
-    # method to return the max value in the linked list
-    # returns the data value and not the node
-    def find_max
-      raise NotImplementedError
+    
+    return min
+  end
+  
+  
+  # method that returns the length of the singly linked list
+  def length
+    node = @head
+    count = 0
+    
+    return 0 if @head == nil
+    
+    until node == nil
+      count += 1
+      node = node.next
     end
-
-    # method to return the min value in the linked list
-    # returns the data value and not the node
-    def find_min
-      raise NotImplementedError
+    return count
+  end
+  
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  def get_at_index(index)
+    return nil if @head == nil
+    current = @head
+    index.times do 
+      current = current.next

```

Comment:
You need to make sure you don't go beyond the bounds of the list.

```suggestion
      return nil if current.nil?
      current = current.next
```